src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,888 @@
+function Base.:^(x::MatrixGroupElem, y::MatrixGroupElem)
+   G = x.parent==y.parent ? x.parent : GL(x.parent.deg, x.parent.ring)
+   if isdefined(x,:X) && isdefined(y,:X) && !(isdefined(x,:elm) && isdefined(y,:elm))
+      return MatrixGroupElem(G, inv(y.X)*x.X*y.X)
+   else
+      return MatrixGroupElem(G,inv(y.elm)*x.elm*y.elm)
+   end
+end
+
+comm(x::MatrixGroupElem, y::MatrixGroupElem) = inv(x)*conj(x,y)
+
+"""
+    det(x::MatrixGroupElem)
+    
+Return the determinant of `x`.
+"""
+det(x::MatrixGroupElem) = det(x.elm)
+
+"""
+    base_ring(x::MatrixGroupElem)
+
+Return the base ring of `x`.
+"""
+base_ring(x::MatrixGroupElem) = x.parent.ring
+
+parent(x::MatrixGroupElem) = x.parent
+
+"""
+    matrix(x::MatrixGroupElem)
+
+Return the underlying `AbstractAlgebra` matrix of `x`.
+"""
+matrix(x::MatrixGroupElem) = x.elm
+
+Base.getindex(x::MatrixGroupElem, i::Int, j::Int) = x.elm[i,j]
+
+"""
+    nrows(x::MatrixGroupElem)
+
+Return the number of rows of the given matrix.
+"""
+nrows(x::MatrixGroupElem) = x.parent.deg
+
+"""
+    trace(x::MatrixGroupElem)
+    tr(x::MatrixGroupElem)
+
+Return the trace of `x`.
+"""
+trace(x::MatrixGroupElem) = trace(x.elm)
+tr(x::MatrixGroupElem) = tr(x.elm)
+
+order(x::MatrixGroupElem) = ZZ(GAP.Globals.Order(x.X))
+order(::Type{T}, x::MatrixGroupElem)  where T<:Number = T(GAP.Globals.Order(x.X))
+
+#FIXME for the following functions, the output may not belong to the parent group of x
+#=
+frobenius(x::MatrixGroupElem, n::Int) = MatrixGroupElem(x.parent, matrix(x.parent.ring, x.parent.deg, x.parent.deg, [frobenius(y,n) for y in x.elm]))
+frobenius(x::MatrixGroupElem) = frobenius(x,1)
+
+transpose(x::MatrixGroupElem) = MatrixGroupElem(x.parent, transpose(x.elm))
+=#
+
+########################################################################
+#
+# Methods on groups
+#
+########################################################################
+
+"""
+    base_ring(G::MatrixGroup)
+
+Return the base ring of the matrix group `G`.
+"""
+base_ring(G::MatrixGroup) = G.ring
+
+"""
+    degree(G::MatrixGroup)
+
+Return the degree of the matrix group `G`, i.e. the number of rows of its matrices.
+"""
+degree(G::MatrixGroup) = G.deg
+
+Base.one(G::MatrixGroup) = MatrixGroupElem(G, identity_matrix(G.ring, G.deg))
+
+function Base.rand(G::MatrixGroup)
+   x_gap = GAP.Globals.Random(G.X)
+   x_oscar = G.mat_iso(x_gap)
+   return MatrixGroupElem(G,x_oscar,x_gap)
+end
+
+function gens(G::MatrixGroup)
+   isdefined(G,:gens) && return G.gens
+   
+   L = GAP.Globals.GeneratorsOfGroup(G.X)
+   G.gens=[MatrixGroupElem(G,G.mat_iso(L[i]),L[i]) for i in 1:length(L)]
+   return G.gens
+end
+
+gen(G::MatrixGroup, i::Int) = gens(G)[i]
+
+Base.getindex(G::MatrixGroup, i::Int) = gen(G, i)
+
+ngens(G::MatrixGroup) = length(gens(G))
+
+function order(G::MatrixGroup)
+   isdefined(G, :order) && return G.order
+
+   G.order = fmpz(BigInt(GAP.Globals.Size(G.X)))
+   return G.order
+end
```

Comment:
This breaks the contract for `order` as it doesn't implement the two argument version. In fact, with my PR #269 it will be sufficient to only implement the two argument version here, say like this:
```suggestion
function order(::Type{T}, G::MatrixGroup) where T
   isdefined(G, :order) || G.order = fmpz(GAP.Globals.Order(G.X))
   return T(G.order)
end
```
However, I still am not convinced that having the `G.order` field is a good idea in the first place; if we drop it, we can also drop this method and just rely on the generic one.

Reply:
It was Claus' suggestion. If I do not have the field `G.order`, I don't have way to store the order other than the underlying GAP object.

Reply:
The line `isdefined(G, :order) || G.order = fmpz(GAP.Globals.Order(G.X))` returns the error `ERROR: syntax: invalid assignment location`. I replaced it by `if !isdefined(G, :order) G.order =  fmpz(GAP.Globals.Order(G.X)) end`.